### PR TITLE
*: Move sending telemetry to Prometheus native functionality

### PR DESCRIPTION
**Pull request by brancz**
This PR moves sending telemetry data to use Prometheus native functionality using remote-write. Beyond the functionality, this changes (and simplifies a lot) the process of adding metrics to telemetry: no generation, no updating in multiple repos - one single source of truth.

This is in proof of concept state. The config is current actually generated from the telemeter repo just for the sake of staying up to date with the current config, the intention is that when this is merged, this repo will be the one and only config. This is targeted for 4.3.

@LiliC @s-urbaniak @paulfantom @squat @metalmatze @kakkoyun @ant31 

cc @smarterclayton @sichvoge

### CODE REVIEW

**Comment by brancz** [comment]
Putting on hold until we've properly tested this.

/hold

**Comment by LiliC** [comment]
@brancz as master-4.3 branch is up to date now, can you change the merge branch to that now, thanks.

**Comment by brancz** [comment]
This PR is ready to be reviewed, but since this data can only be received by the v2 infrastructure, we need to wait until the migration has been fully completed and v1 is shutdown. There is a single test failing in origin, which tested, that the telemeter-client is successfully sending data, which needs to be switched to test that remote-write is successfully doing so. Once the receiving infrastructure is ready, we need to merge https://github.com/openshift/origin/pull/23925 and then this, and then re-enable the successful telemetry sending check.

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1
+data:
+  metrics.yaml: |-
+    "matches":
+    - "{__name__=\"up\"}"
```
Comment (LiliC) [inline manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: So the procedure to add a new item to the allowed list, would be to just modify this ConfigMap?

File: pkg/manifests/config.go [hunk 2]
```diff
@@ -188,7 +189,9 @@ func (c *Config) applyDefaults() {
 		c.HTTPConfig = &HTTPConfig{}
 	}
 	if c.TelemeterClientConfig == nil {
-		c.TelemeterClientConfig = &TelemeterClientConfig{}
+		c.TelemeterClientConfig = &TelemeterClientConfig{
+			TelemeterServerURL: "https://infogw.api.openshift.com/metrics/v1/receive",
```
Comment (LiliC) [inline pkg/manifests/config.go]: Can we not pass this URL via the config options?

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -843,9 +858,72 @@ func (f *Factory) PrometheusK8s(host string) (*monv1.Prometheus, error) {
...
+		spec := monv1.RemoteWriteSpec{
+			URL:         f.config.TelemeterClientConfig.TelemeterServerURL,
+			BearerToken: base64.StdEncoding.EncodeToString(compositeToken),
+			WriteRelabelConfigs: []monv1.RelabelConfig{
+				*selectorRelabelConfig,
+				monv1.RelabelConfig{
+					TargetLabel: "_id",
+					Replacement: f.config.TelemeterClientConfig.ClusterID,
+				},
+				monv1.RelabelConfig{
```
Comment (LiliC) [inline pkg/manifests/manifests.go]: Why is this needed?

File: pkg/relabelgen/relabelgen.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package relabelgen
```
Comment (LiliC) [inline pkg/relabelgen/relabelgen.go]: Missing copyright header.

File: pkg/relabelgen/relabelgen_test.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package relabelgen
```
Comment (LiliC) [inline pkg/relabelgen/relabelgen_test.go]: Missing copyright header.

File: pkg/tasks/prometheus.go [hunk 6]
```diff
@@ -233,8 +233,37 @@ func (t *PrometheusTask) Run() error {
 		}
 	}
 
+	// Create trusted CA bundle ConfigMap.
+	trustedCA, err := t.factory.AlertmanagerTrustedCABundle()
```
Comment (LiliC) [inline pkg/tasks/prometheus.go]: Hmm why the AlertmanagerTrustedCABundle here?

File: pkg/tasks/prometheus.go [hunk 7]
```diff
@@ -233,8 +233,37 @@ func (t *PrometheusTask) Run() error {
...
+	if err != nil {
+		return errors.Wrap(err, "initializing Alertmanager CA bundle ConfigMap failed")
+	}
+
+	trustedCA, err = t.client.CreateIfNotExistConfigMap(trustedCA)
+	if err != nil {
+		return errors.Wrap(err, " creating Alertmanager CA bundle ConfigMap failed")
```
Comment (LiliC) [inline pkg/tasks/prometheus.go]: Alertmanager? You meant prometheus?

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml (hunk 1 above)
Comment (brancz) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: correct

File: pkg/manifests/config.go (hunk 2 above)
Comment (brancz) [reply pkg/manifests/config.go]: this is just the default, for when no option is specified

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (brancz) [reply pkg/manifests/manifests.go]: The main reason is to be able to distinguish between alerts produced in cluster (`alerts`) and alerts produced on the receiving side (`ALERTS`). The naming is a little unfortunate I believe, but we need to stick to this for compatibility for older data.

File: pkg/tasks/prometheus.go (hunk 6 above)
Comment (brancz) [reply pkg/tasks/prometheus.go]: whops, copy pasta

File: pkg/tasks/prometheus.go (hunk 7 above)
Comment (brancz) [reply pkg/tasks/prometheus.go]: yep

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml (hunk 1 above)
Comment (LiliC) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: Do you think we can document this either in the google doc we have or in the README, can be done outside of this PR, just so we don't have to explain to people. :)

Comment (s-urbaniak) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: nit: we should also document the expectations about changes of that configmap in a running cluster, i.e. what happens if somebody (or some system) tries to modify that configmap.

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/manifests/manifests.go]: let's add the above statement as a code comment.

File: pkg/manifests/manifests.go [hunk 8]
```diff
@@ -843,9 +858,72 @@ func (f *Factory) PrometheusK8s(host string) (*monv1.Prometheus, error) {
...
 	if len(f.config.PrometheusK8sConfig.RemoteWrite) > 0 {
 		p.Spec.RemoteWrite = f.config.PrometheusK8sConfig.RemoteWrite
 	}
+	for _, rw := range p.Spec.RemoteWrite {
+		if f.config.HTTPConfig.HTTPProxy != "" {
+			rw.ProxyURL = f.config.HTTPConfig.HTTPProxy
+		}
+		if f.config.HTTPConfig.HTTPSProxy != "" {
+			rw.ProxyURL = f.config.HTTPConfig.HTTPSProxy
+		}
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: afaik the proxy config also has to respect the trusted CA bundle and react to it upon changes. Is this something we can add easily with the existing Prometheus infrastructure without restarting the whole pod?

File: pkg/tasks/prometheus.go [hunk 9]
```diff
@@ -233,8 +233,37 @@ func (t *PrometheusTask) Run() error {
...
+			"prometheus",
+		)
+		if err != nil {
+			return errors.Wrap(err, "deleting old Prometheus configmaps failed")
+		}
+	}
+
 	klog.V(4).Info("initializing Prometheus object")
-	p, err := t.factory.PrometheusK8s(host)
+	p, err := t.factory.PrometheusK8s(host, trustedCA)
```
Comment (s-urbaniak) [inline pkg/tasks/prometheus.go]: ah, here it is! scratch the note above then :-)

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml (hunk 1 above)
Comment (brancz) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: not a nit, I think this is required, will fix

Comment (brancz) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: and once this is merged I will point the current doc at the one in this repo

File: pkg/manifests/manifests.go (hunk 8 above)
Comment (brancz) [reply pkg/manifests/manifests.go]: clarified in https://github.com/openshift/cluster-monitoring-operator/pull/434#discussion_r331925798

File: pkg/relabelgen/relabelgen_test.go (hunk 5 above)
Comment (brancz) [reply pkg/relabelgen/relabelgen_test.go]: fixed

File: pkg/relabelgen/relabelgen.go (hunk 4 above)
Comment (brancz) [reply pkg/relabelgen/relabelgen.go]: fixed

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (brancz) [reply pkg/manifests/manifests.go]: added

**Comment by LiliC** [comment]
You need to rebase here I think :) But otherwise its ready for another review, right?

File: Documentation/data-collection.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
+# OpenShift 4 Data Collection
+
+Red Hat values our customer experience and privacy. It is important to us that our customers understand exactly what we are sending back to engineering and why. During the developer preview or beta release of our software, we want to be able to make changes to our designs and coding practices in real-time based on customer environments. The faster the feedback loop during these development stages the better.
```
Comment (LiliC) [inline Documentation/data-collection.md]: I am not sure this is true only for beta and developer releases, we always want to make sure this works? Or will we stop sending this data at some point?

File: README.md [hunk 11]
```diff
@@ -26,25 +26,12 @@ Metrics are collected from the following components:
 * API server
 * Prometheus (just `prometheus-k8s` for now)
 * Alertmanager
-* Telemeter
 
-## Contributing new component integrations
+## Adding new metrics to be sent via telemetry
 
-The Cluster Monitoring Operator has many builtin `ServiceMonitor` resources which enable discovering the metrics endpoints of a variety of well-known components. Only components that must be created before the cluster monitoring stack belong in this repository, in order to solve the cyclic dependencies of bootstrapping.
+To add new metrics to be sent via telemetry, simply add a selector that matches the time-series to be sent in [manifests/0000_50_cluster_monitoring_operator_04-config.yaml](manifests/0000_50_cluster_monitoring_operator_04-config.yaml).
```
Comment (LiliC) [inline README.md]: This doesn't need any regeneration of the files?

File: cmd/operator/main.go [hunk 12]
```diff
@@ -85,10 +90,30 @@ func Main() int {
...
+	err = yaml.NewYAMLOrJSONDecoder(f, 100).Decode(&telemetryConfig)
+	if err != nil {
+		fmt.Fprint(os.Stderr, "Could not parse telemetry config file.")
+		return 1
+	}
+	f.Close()
+
+	klog.V(4).Info("Configured matches for telemetry:")
+	for _, m := range telemetryConfig.Matches {
+		klog.V(4).Info(m)
```
Comment (LiliC) [inline cmd/operator/main.go]: Hmm so we just log the matches nothing else, how come?

Comment (brancz) [reply cmd/operator/main.go]: These are the selectors, so we essentially log the queries at startup that would match what we're sending. Those are typically referred to as matches/matchers.

File: README.md (hunk 11 above)
Comment (brancz) [reply README.md]: Nope, this will be the only file people will need to modify to add a new metric.

File: Documentation/data-collection.md (hunk 10 above)
Comment (brancz) [reply Documentation/data-collection.md]: I used the existing docs in the telemeter repo for this, so this has been signed off previously, but pinging @sichvoge for extra review.

**Comment by brancz** [comment]
There is an e2e test testing this in origin, which is what's currently failing.

**Comment by paulfantom** [comment]
Since this is removing telemeter-client deployment, do we need to modify `manifests/image-references` in this PR or after we completely remove it in future OpenShift release?

**Comment by brancz** [comment]
I wasn't sure about that one. What I did for now was leave it in as we still have the other telemeter manifests as well, but I don't 100% know the details of removing a component so I might be wrong with that.

File: cmd/operator/main.go [hunk 13]
```diff
@@ -86,10 +91,30 @@ func Main() int {
...
+		return 1
+	}
+
+	telemetryConfig := telemetryConfig{}
+	err = yaml.NewYAMLOrJSONDecoder(f, 100).Decode(&telemetryConfig)
+	if err != nil {
+		fmt.Fprint(os.Stderr, "Could not parse telemetry config file.")
+		return 1
+	}
+	f.Close()
```
Comment (LiliC) [inline cmd/operator/main.go]: This returns an error, can we check it.

File: jsonnet/main.jsonnet [hunk 14]
```diff
@@ -172,6 +172,9 @@ removeLimits(
   { ['prometheus-user-workload/' + name]: kp.prometheusUserWorkload[name] for name in std.objectFields(kp.prometheusUserWorkload) } +
   { ['prometheus-adapter/' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
   { ['grafana/' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) } +
+  // needs to be removed once 4.3 ships, as this is needed for removal of the
```
Comment (LiliC) [inline jsonnet/main.jsonnet]: is this still relevant, as 4.3 sort of is shipped and this is landing in 4.4

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml (hunk 1 above)
Comment (LiliC) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: Did you update this list to reflect the current state in telemeter repo?

File: jsonnet/main.jsonnet (hunk 14 above)
Comment (brancz) [reply jsonnet/main.jsonnet]: Good catch, this should now be 4.4.

File: cmd/operator/main.go (hunk 13 above)
Comment (brancz) [reply cmd/operator/main.go]: Fixed

File: jsonnet/main.jsonnet (hunk 14 above)
Comment (brancz) [reply jsonnet/main.jsonnet]: Fixed.

File: manifests/0000_50_cluster_monitoring_operator_04-config.yaml (hunk 1 above)
Comment (brancz) [reply manifests/0000_50_cluster_monitoring_operator_04-config.yaml]: Currently this is still being generated. As soon as this lands we would remove this being generated, but until then I'd like to keep this to ensure that when this lands this is definitely in sync with what is currently merged in telemeter.

tl;dr yes

**Comment by s-urbaniak** [comment]
@brancz we have a failure in the existing e2e tests:
```
--- FAIL: TestTelemeterTrustedCA (300.03s)
    telemeter_test.go:57: getting trusted CA ConfigMap failed: configmaps "telemeter-trusted-ca-bundle" not found
```

This is a legit failure as I guess this doesn't synthesize `telemeter-trusted-ca-bundle` anymore. I guess we have to redirect it to Prometheus now to assert if the trusted CA bundle has been successfully mounted there, cc'ing @LiliC as we've been adding this e2e test to telemeter before.

**Comment by brancz** [comment]
Ah thanks. Removed that test and trusted CA is already handled for prometheus.

**Comment by LiliC** [comment]
>  I guess we have to redirect it to Prometheus now to assert if the trusted CA bundle has been successfully mounted there

@brancz 
Yes, we should adjust the tests here -> https://github.com/openshift/cluster-monitoring-operator/blob/master/test/e2e/telemeter_test.go#L30

**Comment by s-urbaniak** [comment]
indeeed lgtm! :tada:

**Comment by squat** [comment]
I dreamt about this day 😭
Can we remove the hold?

File: pkg/tasks/telemeter.go [hunk 15]
```diff
@@ -35,134 +35,9 @@ func NewTelemeterClientTask(client *client.Client, factory *manifests.Factory, c
 }
 
 func (t *TelemeterClientTask) Run() error {
-	if t.config.IsEnabled() {
-		return t.create()
-	}
-
 	return t.destroy()
```
Comment (s-urbaniak) [inline pkg/tasks/telemeter.go]: just a very small nit (we can do it also as a follow-up). let's document why we are "just" destroying telemeter client here, namely to catch the upgrade path to remote-write. it could be non-obvious when someone looks in here only.

**Comment by s-urbaniak** [comment]
yay `ci/prow/e2e-aws-operator` is green!

**Comment by brancz** [comment]
/hold

Now comes the time that we need to do the dance with https://github.com/openshift/origin/pull/23925.

**Comment by brancz** [comment]
There seems to be a legitimately firing alert. I’ll look deeper into it.

**Comment by LiliC** [comment]
@brancz feel free to remove hold whenever 🎉 lets merge this!! :)
